app/scripts/migrations/081.js

```diff
@@ -0,0 +1,68 @@
+import { cloneDeep } from 'lodash';
+import {
+  WALLET_SNAP_PERMISSION_KEY,
+  SnapCaveatType,
+} from '@metamask/rpc-methods';
+
+const version = 81;
+
+/**
+ * Prior to this migration, snap <> dapp permissions were wildcards i.e. `wallet_snap_*`.
+ * Now the permission has been changed to `wallet_snap` and the current snap permissions
+ * that are under wildcards will be added as caveats to a parent `wallet_snap` permission.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    const newState = transformState(state);
+    versionedData.data = newState;
+    return versionedData;
+  },
+};
+
+function transformState(state) {
+  const PermissionController = state?.PermissionController || {};
+
+  const { subjects } = PermissionController;
+
+  const snapPrefix = 'wallet_snap_';
+
+  for (const [subjectName, subject] of Object.entries(subjects)) {
+    // we keep track of the latest permission's date and associated id
+    // to assign to the wallet_snap permission
+    let date = 1;
+    let id;
+    const { permissions } = subject;
+    const updatedPermissions = { ...permissions };
+    for (const [permissionName, permission] of Object.entries(permissions)) {
```

Comment:
I think this migration could be written in a way that makes it more obvious what is going on.

I would prefer if we did something along the lines of:
1. Find all snaps and turn them into caveats on one object
2. Delete old snap keys from the subject permissions
3. Add permission object from 1

Reply:
I think we should also get a second opinion from someone who is experienced with the migrations since this is a more complex one.

Reply:
Breaking up things to make it readable will likely introduce extra iteration. This is not part of the codebase that will see a lot of traffic, so I'm fine keeping it this way, I did however add extensive comments.